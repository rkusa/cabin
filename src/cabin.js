/**
 * @param {number} eventId
 * @param {object | URLSearchParams} payload
 * @param {Node} target
 * @param {AbortController | undefined} abortController
 * @param {WeakMap<HTMLElement, bool> | undefined} disabledBefore
 */
async function update(
  eventId,
  payload,
  target,
  abortController,
  disabledBefore,
) {
  if (this.abortController) {
    this.abortController.abort();
  }
  abortController = this.abortController =
    abortController ?? new AbortController();
  const signal = this.abortController.signal;
  if (signal.aborted) {
    return;
  }

  try {
    let state = undefined;
    if (
      target.firstChild &&
      target.firstChild instanceof HTMLScriptElement &&
      target.firstChild.getAttribute("type") === "application/json"
    ) {
      state = target.firstChild.innerText;
    }

    /** @type {RequestInit} */
    const req =
      payload instanceof URLSearchParams
        ? {
            signal,
            method: "PUT",
            headers: {
              "x-cabin": "boundary",
            },
            body: (() => {
              const formData = new FormData();
              formData.append(
                "payload",
                new Blob([payload.toString()], {
                  type: "application/x-www-form-urlencoded",
                }),
              );
              formData.append("event_id", eventId);
              if (state) {
                formData.append(
                  "state",
                  new Blob([state], { type: "application/json" }),
                );
              }
              return formData;
            })(),
          }
        : {
            signal,
            method: "PUT",
            headers: {
              "Content-Type": "application/json",
              "x-cabin": "boundary",
            },
            body: `{"eventId":${eventId},"payload":${JSON.stringify(payload)}${
              state ? `,"state":${state}` : ""
            }}`,
          };
    const endpoint =
      target instanceof CabinBoundary
        ? `/__boundary/${target.getAttribute("name")}`
        : location.href;
    const res = await fetch(endpoint, req);
    if (signal.aborted) {
      return;
    }

    const url = new URL(res.url);
    if (res.ok && res.redirected && url.pathname === "/client_redirect") {
      window.location = url.search.substring(1);
      return;
    }

    if (!target.parentNode) {
      return;
    }

    if (res.status !== 200) {
      throw new Error(`received unexpected status code: ${res.status}`);
    }

    const html = await res.text();

    console.time("patch");
    const template = document.createElement("template");
    template.innerHTML = html;
    patchChildren(target, template.content, {}, disabledBefore);
    console.timeEnd("patch");

    const rewriteUrl = res.headers.get("location");
    if (rewriteUrl && `${location.pathname}${location.search}` !== rewriteUrl) {
      history.pushState(null, undefined, rewriteUrl);
    }

    const newTitle = res.headers.get("x-cabin-title");
    if (newTitle) {
      document.title = newTitle;
    }

    // TODO: prevent endless event loops
    {
      const eventId = res.headers.get("x-cabin-event");
      const payload = res.headers.get("x-cabin-payload");
      if (eventId && payload) {
        target.dispatchEvent(
          new CustomEvent("cabinFire", {
            detail: { eventId, payload: JSON.parse(payload) },
            bubbles: true,
          }),
        );
      }
    }
  } catch (err) {
    if (err instanceof DOMException && err.name === "AbortError") {
      // ignore
    } else {
      throw err;
    }
  } finally {
    if (this.abortController === abortController) {
      this.abortController = undefined;
    }
  }
}

function setUpEventListener(el, eventName, opts) {
  const attrName = `cabin-${eventName}`;
  /** @type {WeakMap<Element, AbortController>} */
  const abortControllers = opts.abortControllers ?? new WeakMap();

  /**
   * @this {HTMLElement}
   * @param {Event} e
   */
  async function handleEvent(e) {
    /** @type {Element} */
    let node = e.target;

    do {
      const eventId = e.detail?.eventId ?? node.getAttribute(attrName);
      if (!eventId) {
        continue;
      }

      // The boundary only intercepts certain events
      if (opts.events && !opts.events.has(eventId)) {
        return;
      }

      // The internal state/view of the boundary is possibly going to change due to this event. To
      // force an update for the boundary if its parent view changes, remove all hash attributes
      // from ascendents up until the next cabin boundary.
      if (el !== document) {
        let el = this;
        do {
          el.removeAttribute("hash");
        } while ((el = el.parentElement) && !(el instanceof CabinBoundary));
      }

      if (opts.disable && node.disabled) {
        return;
      }

      // Force refresh everything from the target to its boundary to ensure that the server might
      // override e.g. input states.
      if (opts.dirty) {
        let el = node;
        do {
          el.removeAttribute("hash");
        } while ((el = el.parentElement) && !(el instanceof CabinBoundary));
      }

      {
        const abortController = abortControllers.get(node);
        if (abortController) {
          abortController.abort();
        }
      }

      const abortController = new AbortController();
      abortControllers.set(node, abortController);

      const isSubmitEvent = eventName === "submit";
      e.stopPropagation();
      if (
        (opts.preventDefault && !(node instanceof HTMLInputElement)) ||
        isSubmitEvent
      ) {
        e.preventDefault();
      }

      if (opts.debounce) {
        await new Promise((resolve) => setTimeout(resolve, opts.debounce));
        if (abortController.signal.aborted) {
          return;
        }
      }

      /** @type {WeakMap<HTMLElement, bool>} */
      const disabledBefore = new WeakMap();

      try {
        let payload;
        if (isSubmitEvent) {
          payload = new URLSearchParams(new FormData(node));
        } else if (opts?.eventPayload) {
          payload = JSON.parse(
            Object.entries(opts.eventPayload(e)).reduce(
              (result, [placeholder, value]) =>
                result.replace(placeholder, value),
              node.getAttribute(`${attrName}-payload`),
            ),
          );
        } else {
          payload =
            e.detail && typeof e.detail === "object" && "payload" in e.detail
              ? e.detail.payload
              : JSON.parse(node.getAttribute(`${attrName}-payload`));
        }

        if (isSubmitEvent) {
          // disable whole form
          for (const el of node.elements) {
            disabledBefore.set(el, el.disabled);
            el.disabled = true;
          }
        } else if (opts.disable) {
          disabledBefore.set(node, node.disabled);
          node.disabled = true;
        }

        // Check for, and if exists, apply pre-rendered instances of this boundary
        if (!isSubmitEvent && this instanceof CabinBoundary) {
          let templates = [];
          let template = this.lastElementChild;
          while (
            template &&
            template instanceof HTMLTemplateElement &&
            template.hasAttribute("event-id") &&
            template.hasAttribute("event-payload")
          ) {
            templates.push(template);
            template = template.previousElementSibling;
          }
          // TODO: might not always be the same, somehow use same string representation of the
          // payload as the server?
          const payloadStr = JSON.stringify(payload);
          for (const template of templates) {
            if (
              template.getAttribute("event-id") === eventId &&
              template.getAttribute("event-payload") === payloadStr
            ) {
              console.time("patch");
              patchChildren(el, template.content, {}, disabledBefore);
              // put back prerendered templates
              for (const template of templates) {
                this.appendChild(template);
              }
              console.timeEnd("patch");
              return;
            }
          }
        }

        await update(
          parseInt(eventId),
          payload,
          el == document ? document.body : el,
          abortController,
          disabledBefore,
        );
      } catch (err) {
        throw err;
      } finally {
        if (isSubmitEvent) {
          // restore disabled state
          for (const el of node.elements) {
            const before = disabledBefore.get(el);
            if (before !== undefined) {
              el.disabled = before;
            }
          }
        } else if (opts.disable) {
          node.disabled = disabledBefore.get(node) ?? false;
        }

        abortControllers.delete(node);
      }

      break;
    } while ((node = node.parentElement));
  }

  el.addEventListener(eventName, function (e) {
    handleEvent.call(this, e).catch((err) => {
      console.error(err);
    });
  });
}

/**
 * @param {Node} rootBefore
 * @param {Node} rootAfter
 * @param {Record<string, Node>} orphanKeyed
 * @param {WeakMap<HTMLElement, bool> | undefined} disabledBefore
 */
function patchChildren(rootBefore, rootAfter, orphanKeyed, disabledBefore) {
  // console.log("apply", rootBefore, rootAfter);

  let nodeBefore = rootBefore.firstChild;
  let nodeAfter = rootAfter.firstChild;

  // Neither node has children, done here
  if (!nodeBefore && !nodeAfter) {
    return;
  }

  /** @type {Node | null} */
  let nextBefore = null;
  /** @type {Node | null} */
  let nextAfter = null;

  do {
    nextBefore = null;
    nextAfter = null;
    // console.log(nodeBefore, "vs", nodeAfter);

    if (!nodeAfter) {
      // console.log("removed", nodeBefore);
      nextBefore = nodeBefore.nextSibling;
      rootBefore.removeChild(nodeBefore);
      continue;
    }

    // This node and all its next siblings are new nodes and can be directly added
    if (nodeBefore === null) {
      // console.log("append new", nodeAfter, "and siblings");
      const fragment = document.createDocumentFragment();
      let node = nodeAfter;
      while (node) {
        let next = node.nextSibling;
        if (isKeyedElement(node)) {
          // Only checking orphan nodes as there are no siblings remaining to check anyway
          const previous = orphanKeyed[node.id];
          if (previous) {
            // console.log(`found existing ${node.id} and moved it into place`);
            fragment.appendChild(previous);
            delete orphanKeyed[node.id];
            node = next;
            continue;
          }
        }
        fragment.appendChild(node);
        node = next;
      }

      rootBefore.appendChild(fragment);
      return;
    }

    // re-use if found somewhere else in the tree
    if (
      isKeyedElement(nodeAfter) &&
      (!isKeyedElement(nodeBefore) || nodeBefore.id !== nodeAfter.id)
    ) {
      const previous =
        document.getElementById(nodeAfter.id) ?? orphanKeyed[nodeAfter.id];
      nextBefore = nodeBefore;
      nextAfter = nodeAfter.nextSibling;
      if (
        previous &&
        (!previous.parentNode || previous.parentNode == nodeBefore.parentNode)
      ) {
        // console.log(`found existing ${nodeAfter.id} and moved it into place`);
        rootBefore.insertBefore(previous, nodeBefore);
        delete orphanKeyed[nodeAfter.id];
        nodeBefore = previous;
      } else {
        // console.log("new iter item, move new into place");
        rootBefore.insertBefore(nodeAfter, nodeBefore);
        continue;
      }
    }

    // type changed, replace completely
    else if (
      nodeBefore.nodeType !== nodeAfter.nodeType ||
      nodeBefore.nodeName !== nodeAfter.nodeName
    ) {
      // console.log("replace due to type change");
      nextBefore = nodeBefore.nextSibling;
      nextAfter = nodeAfter.nextSibling;
      rootBefore.replaceChild(nodeAfter, nodeBefore);

      // Keep it around in case it got moved
      if (isKeyedElement(nodeBefore)) {
        orphanKeyed[nodeBefore.id] = nodeBefore;
      }

      continue;
    }

    switch (nodeAfter.nodeType) {
      case Node.COMMENT_NODE:
        throw new Error("unexpected comment");

      case Node.ELEMENT_NODE:
        // skip sub-tree if hash is unchanged
        if (
          nodeBefore.hasAttribute("hash") &&
          nodeAfter.hasAttribute("hash") &&
          nodeBefore.getAttribute("hash") == nodeAfter.getAttribute("hash")
        ) {
          // console.log("skip due to unchanged hash");
          break;
        }

        // console.log("patch attributes");
        patchAttributes(nodeBefore, nodeAfter, disabledBefore);
        patchChildren(nodeBefore, nodeAfter, orphanKeyed, disabledBefore);
        break;

      case Node.TEXT_NODE:
        if (nodeAfter.textContent !== nodeBefore.textContent) {
          // console.log("update text");
          nodeBefore.textContent = nodeAfter.textContent;
        } else {
          // console.log("text is unchanged");
        }
        break;
    }
  } while (
    (nextAfter !== null || nextBefore !== null
      ? ((nodeAfter = nextAfter), (nodeBefore = nextBefore))
      : ((nodeAfter = nodeAfter?.nextSibling),
        (nodeBefore = nodeBefore?.nextSibling)),
    nodeAfter || nodeBefore)
  );
}

/**
 * @param {Element} childBefore
 * @param {Element} childAfter
 * @param {WeakMap<HTMLElement, bool> | undefined} disabledBefore
 */
function patchAttributes(childBefore, childAfter, disabledBefore) {
  // special handling for certain elements
  switch (childAfter.nodeName) {
    case "DIALOG":
      if (childAfter.hasAttribute("open")) {
        childBefore.show();
      } else {
        childBefore.close();
      }
      childAfter.removeAttribute("open");
  }

  const oldAttributeNames = new Set(childBefore.getAttributeNames());
  if (childBefore instanceof HTMLInputElement) {
    oldAttributeNames.add("checked");
    oldAttributeNames.add("selected");
    oldAttributeNames.add("value");
  }
  if (
    childBefore instanceof HTMLInputElement ||
    childBefore instanceof HTMLTextAreaElement
  ) {
    oldAttributeNames.add("disabled");
  }

  for (const name of childAfter.getAttributeNames()) {
    oldAttributeNames.delete(name);

    if (ignoreAttribute(childAfter, name)) {
      continue;
    }

    const newValue = childAfter.getAttribute(name);
    switch (name) {
      case "value":
        if (childBefore.value !== newValue) {
          // console.log("update attribute", name);
          childBefore.value = newValue;
        }
        break;
      case "checked":
        if (childBefore.checked !== childAfter.checked) {
          // console.log("update attribute", name);
          childBefore.checked = childAfter.checked;
        }
        break;
      case "selected":
        if (childBefore.selected !== childAfter.selected) {
          // console.log("update attribute", name);
          childBefore.selected = childAfter.selected;
        }
        break;
      case "disabled":
        disabledBefore?.set(childBefore, childAfter.disabled);
      // fallthrough
      default:
        if (childBefore.getAttribute(name) !== newValue) {
          // console.log("update attribute", name);
          childBefore.setAttribute(name, newValue);
        }
        break;
    }
  }

  // delete attributes that are not set anymore
  for (const name of oldAttributeNames) {
    if (ignoreAttribute(childBefore, name)) {
      continue;
    }
    // console.log("remove attribute", name);
    switch (name) {
      case "value":
        childBefore.value = "";
        break;
      case "checked":
        childBefore.checked = false;
        break;
      case "selected":
        childBefore.selected = false;
        break;
      case "disabled":
        childBefore.disabled = false;
        break;
      default:
        childBefore.removeAttribute(name);
        break;
    }
  }
}

/**
 * @param {Element} el
 * @param {string} attr
 */
function ignoreAttribute(el, attr) {
  switch (el.nodeName) {
    case "DIALOG":
      return attr === "open";
    default:
      return false;
  }
}

/**
 * @param {Node} node
 * @return {boolean}
 */
function isKeyedElement(node) {
  return node.nodeType === Node.ELEMENT_NODE && node.nodeName === "CABIN-KEYED";
}

function setupEventListeners(el) {
  let events =
    el instanceof CabinBoundary
      ? new Set(
          el
            .getAttribute("events")
            .split(",")
            .filter((s) => s.length > 0),
        )
      : null;
  if (events && events.size === 0) {
    events = undefined;
  }

  setUpEventListener(el, "click", {
    events,
    preventDefault: true,
    disable: true,
  });
  setUpEventListener(el, "transitionend", {
    events,
  });
  setUpEventListener(el, "animationend", {
    events,
  });
  setUpEventListener(el, "change", {
    events,
    eventPayload: (e) => ({
      "_##InputValue": e.target.value,
      '"_##InputChecked"': e.target.checked,
    }),
    dirty: true,
  });
  setUpEventListener(el, "input", {
    events,
    debounce: 500,
    eventPayload: (e) => ({
      "_##InputValue": e.target.value,
      '"_##InputChecked"': e.target.checked,
    }),
    dirty: true,
  });
  setUpEventListener(el, "submit", {
    events,
    preventDefault: true,
  });
  setUpEventListener(el, "cabinFire", {
    events,
  });
}

class CabinBoundary extends HTMLElement {
  constructor() {
    super();

    setupEventListeners(this);
  }
}

customElements.define("cabin-boundary", CabinBoundary);

setupEventListeners(document);

document.addEventListener("cabinRefresh", async function () {
  // Force update all boundary content
  for (let el of document.querySelectorAll("cabin-boundary")) {
    do {
      el.removeAttribute("hash");
    } while ((el = el.parentElement) && !(el instanceof CabinBoundary));
  }
  await update(0, {}, document.body);
});

document.addEventListener("cabinFire", async function (e) {
  await update(e.detail?.eventId, e.detail?.payload, document.body);
});

window.addEventListener("popstate", () => {
  document.dispatchEvent(new CustomEvent("cabinRefresh"));
});

window.addEventListener("pageshow", function (e) {
  // if loaded from cache, refresh page data
  if (e.persisted) {
    document.dispatchEvent(new CustomEvent("cabinRefresh"));
  }
});
