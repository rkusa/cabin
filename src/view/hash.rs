use std::hash::Hasher;

use serde::{Deserialize, Serialize};
use twox_hash::XxHash32;

pub struct HashTree {
    hasher: XxHash32,
    // Represent the tree as a flat structure to reduce allocations (as otherwise each level
    // would require a new allocation for its own vec).
    tree: Vec<Marker>,
    previous_tree: Option<Vec<Marker>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ViewHashTree(Vec<Marker>);

#[derive(Debug, PartialEq, Eq)]
enum Marker {
    Start,
    End(u32),
}

pub struct Node<'a> {
    parent_hasher: XxHash32,
    hash_tree: &'a mut HashTree,
}

impl Default for HashTree {
    fn default() -> Self {
        Self {
            hasher: XxHash32::default(),
            tree: Vec::with_capacity(32),
            previous_tree: None,
        }
    }
}

impl HashTree {
    pub fn from_previous_tree(previous_tree: ViewHashTree) -> Self {
        Self {
            previous_tree: Some(previous_tree.0),
            ..Default::default()
        }
    }

    pub fn hash(&self) -> u32 {
        self.hasher.finish() as u32
    }

    pub fn node(&mut self) -> Node<'_> {
        self.tree.push(Marker::Start);
        // When entering a node, set the hasher of the [HashTree] to a fresh hasher and store its
        // current hasher in the node. It is swapped back once the node gets dropped.
        let parent_hasher = std::mem::take(&mut self.hasher);
        Node {
            parent_hasher,
            hash_tree: self,
        }
    }

    pub fn finish(mut self) -> ViewHashTree {
        self.tree.push(Marker::End(self.hash()));
        ViewHashTree(self.tree)
    }

    pub fn changed_or_else<R>(&self, hash: u32, f: impl FnOnce() -> R) -> Option<R> {
        if let Some(Marker::End(previous)) = self
            .previous_tree
            .as_ref()
            .and_then(|t| t.get(self.tree.len() - 1))
        {
            if *previous == hash {
                return None;
            }
        }
        Some(f())
    }
}

impl<'a> Node<'a> {
    pub fn content(&mut self) -> &mut HashTree {
        self.hash_tree
    }

    pub fn end(self) -> u32 {
        self.hash_tree.hash()
    }
}

impl Hasher for HashTree {
    fn finish(&self) -> u64 {
        self.hasher.finish()
    }

    fn write(&mut self, bytes: &[u8]) {
        self.hasher.write(bytes);
    }
}

impl<'a> Hasher for Node<'a> {
    fn finish(&self) -> u64 {
        self.hash_tree.finish()
    }

    fn write(&mut self, bytes: &[u8]) {
        self.hash_tree.write(bytes);
    }
}

impl<'a> Drop for Node<'a> {
    fn drop(&mut self) {
        let hash = self.hash_tree.hash();
        self.hash_tree.tree.push(Marker::End(hash));
        self.parent_hasher.write_u32(hash);
        std::mem::swap(&mut self.hash_tree.hasher, &mut self.parent_hasher);
    }
}

impl Serialize for Marker {
    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            Marker::Start => s.serialize_str("s"),
            Marker::End(hash) => s.serialize_u32(*hash),
        }
    }
}

mod deserialize {
    use std::fmt;

    use serde::de::{self, Visitor};
    use serde::Deserialize;

    use super::Marker;

    struct MarkerVisitor;

    impl<'de> Visitor<'de> for MarkerVisitor {
        type Value = Marker;

        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
            formatter.write_str(r#"an u32 or "s""#)
        }

        fn visit_i8<E>(self, v: i8) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            self.visit_i32(v as i32)
        }

        fn visit_i16<E>(self, v: i16) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            self.visit_i32(v as i32)
        }

        fn visit_i32<E>(self, v: i32) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(Marker::End(v.try_into().map_err(de::Error::custom)?))
        }

        fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(Marker::End(v.try_into().map_err(de::Error::custom)?))
        }

        fn visit_u8<E>(self, v: u8) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            self.visit_u32(v as u32)
        }

        fn visit_u16<E>(self, v: u16) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            self.visit_u32(v as u32)
        }

        fn visit_u32<E>(self, v: u32) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(Marker::End(v))
        }

        fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(Marker::End(v.try_into().map_err(de::Error::custom)?))
        }

        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            if v == "s" {
                Ok(Marker::Start)
            } else {
                Err(de::Error::invalid_type(de::Unexpected::Str(v), &self))
            }
        }
    }

    impl<'de> Deserialize<'de> for Marker {
        fn deserialize<D>(d: D) -> Result<Self, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            d.deserialize_any(MarkerVisitor)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{html, Render, View};

    fn hash(tag: &str) -> u32 {
        let mut hasher = XxHash32::default();
        hasher.write(tag.as_bytes());
        hasher.finish() as u32
    }

    #[test]
    fn test_consistent_hash() {
        let mut hash_tree = HashTree::default();
        hash_tree.write(b"div");
        assert_eq!(hash_tree.hash(), 3201766860);
        assert_eq!(hash_tree.hash(), 3201766860);
    }

    #[test]
    fn test_hash_tree_tuple_flattened() {
        fn component() -> impl View {
            (html::div(()), (html::div(()), html::div(())))
        }
        let mut hash_tree = HashTree::default();
        let renderer = component().render(&mut hash_tree).unwrap();
        let mut out = String::new();
        renderer.render(&mut out, false).unwrap();
        assert_eq!(
            hash_tree.tree,
            vec![
                Marker::Start,
                Marker::End(hash("div")),
                Marker::Start,
                Marker::End(hash("div")),
                Marker::Start,
                Marker::End(hash("div"))
            ]
        );
    }

    #[test]
    fn test_hash_tree_nested() {
        fn component() -> impl View {
            (html::div(()), html::div(("foobar", html::div(()))))
        }
        let mut hash_tree = HashTree::default();
        let renderer = component().render(&mut hash_tree).unwrap();
        let mut out = String::new();
        renderer.render(&mut out, false).unwrap();
        let expected_parent_hash = {
            let mut hasher = XxHash32::default();
            hasher.write(b"div");
            hasher.write_u32(hash("foobar"));
            hasher.write_u32(hash("div"));
            hasher.finish() as u32
        };
        assert_eq!(
            hash_tree.tree,
            vec![
                Marker::Start,
                Marker::End(hash("div")),
                Marker::Start,
                Marker::Start,
                Marker::End(hash("foobar")),
                Marker::Start,
                Marker::End(hash("div")),
                Marker::End(expected_parent_hash)
            ]
        );
    }

    #[test]
    fn test_serde() {
        let hash_tree = vec![
            Marker::Start,
            Marker::End(1),
            Marker::Start,
            Marker::Start,
            Marker::End(2),
            Marker::Start,
            Marker::End(3),
            Marker::End(4),
            Marker::End(5),
        ];
        let serialized = serde_json::to_string(&hash_tree).unwrap();
        assert_eq!(serialized, r#"["s",1,"s","s",2,"s",3,4,5]"#);
        let deserialized: Vec<Marker> = serde_json::from_str(&serialized).unwrap();
        assert_eq!(deserialized, hash_tree);
    }
}
