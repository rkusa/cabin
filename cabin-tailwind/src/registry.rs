use std::borrow::Cow;
use std::collections::HashMap;
use std::collections::hash_map::DefaultHasher;
use std::fmt::{self, Write};
use std::hash::Hasher;

use bytes::Bytes;
use cabin::View;
use twox_hash::XxHash32;

use super::Utility;

type Order = (usize, usize, u32);

pub struct StyleRegistry {
    classes: HashMap<String, Order>,
}

impl StyleRegistry {
    pub fn with(mut self, styles: &[fn(&mut StyleRegistry)]) -> Self {
        for f in styles {
            (f)(&mut self);
        }
        self
    }

    pub fn add(&mut self, major_order: usize, styles: &[&dyn Utility]) -> String {
        let mut sorted = styles
            .iter()
            .map(|s| (hash_style(*s), *s))
            .collect::<Vec<_>>();
        sorted.sort_by_key(|(hash, _)| *hash);

        let grouped = sorted.into_iter().map(|(_, s)| s).fold(
            HashMap::<_, Vec<_>>::new(),
            |mut grouped, style| {
                let mut hasher = DefaultHasher::new();
                style.hash_modifier(&mut hasher);
                let hash = hasher.finish();
                grouped.entry(hash).or_default().push(style);
                grouped
            },
        );

        // As everything is written to a string, all unwraps below are fine.
        let mut all_names = String::with_capacity(8);
        for (_, mut styles) in grouped {
            styles.sort_by_key(|s| s.order());

            let mut out = String::with_capacity(256);
            let pos = out.len();

            writeln!(&mut out, "@keyframes ").unwrap();
            let animation_name_offset1 = out.len();
            write!(&mut out, "          {{").unwrap();
            writeln!(&mut out, "  from {{").unwrap();
            let before_animate_from = out.len();
            for style in &styles {
                style.write_animate_from(&mut out).unwrap();
            }
            let has_animate_from = out.len() > before_animate_from;
            writeln!(&mut out, "  }}").unwrap();
            writeln!(&mut out, "  to {{").unwrap();
            let before_animate_to = out.len();
            for style in &styles {
                style.write_animate_to(&mut out).unwrap();
            }
            let has_animate_to = out.len() > before_animate_to;
            writeln!(&mut out, "  }}").unwrap();
            writeln!(&mut out, "}}").unwrap();

            let has_animation = has_animate_from || has_animate_to;
            if !has_animation {
                out.truncate(pos);
            }

            // already grouped by variants, so just writing it once (from the first), is enough
            if let Some(style) = styles.first() {
                style.selector_prefix(&mut out).unwrap();
            }
            let class_name_offset = out.len();
            write!(&mut out, "          ").unwrap();
            // already grouped by variants, so just writing it once (from the first), is enough
            if let Some(style) = styles.first() {
                style.selector_suffix(&mut out).unwrap();
            }
            writeln!(&mut out, " {{").unwrap();
            let mut animation_name_offset2 = 0;
            if has_animation {
                // TODO: make easing function, delay, duration, etc. customizable
                write!(&mut out, "animation: 250ms ease-in-out 1 forwards ").unwrap();
                animation_name_offset2 = out.len();
                writeln!(&mut out, "         ;").unwrap();
            }

            // already grouped by variants, so just writing it once (from the first), is enough
            if let Some(style) = styles.first() {
                style.selector_declarations(&mut out).unwrap();
            }
            for style in &styles {
                style.declarations(&mut out).unwrap();
            }
            write!(&mut out, "}}").unwrap();
            if let Some(style) = styles.first() {
                style.suffix(&mut out).unwrap();
            }
            writeln!(&mut out).unwrap();

            let hash = XxHash32::oneshot(0, out[pos..].as_bytes());

            // write actual class name, prepend `_` as it class names must not start with a number
            let name = styles
                .first()
                .and_then(|s| s.override_class_name().map(Cow::Borrowed))
                .unwrap_or_else(|| Cow::Owned(format!("_{hash:x}")));

            let offset = class_name_offset + 9 - name.len();
            out.replace_range(offset..offset + 1, ".");
            out.replace_range(offset + 1..offset + 1 + name.len(), &name);

            if has_animation {
                let offset = animation_name_offset1 + 9 - name.len();
                out.replace_range(offset..offset + name.len(), &name);
                let offset = animation_name_offset2 + 9 - name.len();
                out.replace_range(offset..offset + name.len(), &name);
            }

            let minor_order = styles.iter().map(|s| s.order()).max().unwrap_or_default();
            self.classes.insert(out, (major_order, minor_order, hash));

            if !all_names.is_empty() {
                all_names.push(' ');
            }
            all_names.push_str(&name);
        }

        all_names
    }

    #[cfg(not(target_arch = "wasm32"))]
    pub fn build(self, include_base: bool) -> StyleSheet {
        let mut style_sheet = self.classes.into_iter().collect::<Vec<_>>();
        style_sheet.sort_by_key(|(_, (o1, o2, h))| (*o1, *o2, *h));

        let other = [
            #[cfg(not(test))]
            include_str!("./base.css"),
            #[cfg(all(feature = "preflight", not(test)))]
            include_str!("./preflight/preflight-v3.2.4.css"),
            #[cfg(all(feature = "forms", not(test)))]
            include_str!("./forms/forms-v0.5.3.css"),
        ];

        let css: String = if include_base {
            other
                .into_iter()
                .map(Cow::Borrowed)
                .chain(style_sheet.into_iter().map(|(s, _)| Cow::Owned(s)))
                .collect()
        } else {
            style_sheet.into_iter().map(|(s, _)| s).collect()
        };
        let hash = cabin::content_hash(css.as_bytes());
        StyleSheet {
            content: Bytes::from(css),
            path: format!("/styles.css?{hash}"),
        }
    }
}

impl Default for StyleRegistry {
    fn default() -> Self {
        Self {
            classes: HashMap::with_capacity(256),
        }
    }
}

pub struct StyleSheet {
    pub content: Bytes,
    pub path: String,
}

impl StyleSheet {
    pub fn link(&'static self) -> impl View {
        use cabin::html;
        use cabin::html::Common;
        use html::elements::link::Link;

        html::link()
            .id("cabin-styles")
            .rel(html::elements::link::Rel::StyleSheet)
            .href(&self.path)
    }
}

fn hash_style(style: &dyn Utility) -> u64 {
    struct HashWriter(DefaultHasher);

    impl fmt::Write for HashWriter {
        fn write_str(&mut self, s: &str) -> fmt::Result {
            self.0.write(s.as_bytes());
            Ok(())
        }
    }

    let mut writer = HashWriter(DefaultHasher::default());
    style.declarations(&mut writer).ok();
    style.hash_modifier(&mut writer.0);
    writer.0.finish()
}

#[test]
fn test_deduplication() {
    // Generate same class name if styles are the same just in a different order.

    use crate::utilities::{BLOCK, p};

    let mut r = StyleRegistry {
        classes: Default::default(),
    };
    let a = r.add(0, &[&BLOCK, &p(4)]);
    let b = r.add(0, &[&p(4), &BLOCK]);
    assert_eq!(a, b);
    insta::assert_snapshot!(std::str::from_utf8(&r.build(true).content).unwrap());
}

#[test]
fn test_order() {
    // Test order of @media statements

    use super::Responsive;
    use crate::utilities::BLOCK;

    let mut r = StyleRegistry {
        classes: Default::default(),
    };
    r.add(
        0,
        &[
            &BLOCK.sm().max_md(),
            &BLOCK.md(),
            &BLOCK.max_sm(),
            &BLOCK.max_md(),
            &BLOCK.sm(),
        ],
    );
    insta::assert_snapshot!(std::str::from_utf8(&r.build(true).content).unwrap());
}
