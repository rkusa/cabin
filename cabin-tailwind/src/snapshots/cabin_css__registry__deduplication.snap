---
source: cabin-tailwind/src/registry.rs
expression: r.out
---
._fa8ecfbb {
display: block;
padding: 1rem;
}
